File: w2/db_query/backend/src/db_query/schemas/query.py
"""Pydantic models for ``POST /api/v1/dbs/{name}/query`` (camelCase JSON)."""

from typing import Any

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel


class CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class QueryRequest(CamelModel):
    sql: str


class QueryResult(CamelModel):
    columns: list[str]
    rows: list[list[Any]]
    truncated: bool
    max_rows: int

File: w2/db_query/frontend/src/api/query.ts
/** POST `/api/v1/dbs/{name}/query` */

const base =
  (import.meta.env.VITE_API_BASE_URL ?? 'http://127.0.0.1:8000').replace(/\/$/, '')

async function parseJson<T>(res: Response): Promise<T> {
  const text = await res.text()
  let data: unknown
  try {
    data = text ? JSON.parse(text) : null
  } catch {
    throw new Error(`Invalid JSON (${res.status})`)
  }
  if (!res.ok) {
    const err = data as { error?: { message?: string } } | null
    const msg = err?.error?.message ?? `HTTP ${res.status}`
    throw new Error(msg)
  }
  return data as T
}

export type QueryResult = {
  columns: string[]
  rows: unknown[][]
  truncated: boolean
  maxRows: number
}

export async function executeQuery(dbName: string, sql: string): Promise<QueryResult> {
  const res = await fetch(
    `${base}/api/v1/dbs/${encodeURIComponent(dbName)}/query`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ sql }),
    },
  )
  return parseJson<QueryResult>(res)
}
